### PR TITLE
Separate text encoder LR; logging & pbar improvements; misc sample generator tweaks

### PR DESCRIPTION
**Pull request by damian0815**
* Implement different LR for the text encoder. Control with `text_encoder_lr_scale` in `optimizer.json` (default=1).

i also started making some small tweaks to the sample generator and it kinda got away on me:

* Tidy up validation and sample generator progress bars by properly passing `position` and `leave=False` arguments to `tqdm`.
* Add `aspect_ratio` feature to samples config, to automatically determine an appropriate aspect ratio for the training size. 
* Add `generate_samples_every_n_steps` and `generate_pretrain_samples` to samples config file.
* Improve sample generation logging.
* Update docs

### CODE REVIEW

**Comment by damian0815** [comment]
it'd be easy enough to split off the text encoder LR stuff from the sample generator/logging/pbar tweaks if you'd prefer it that way.

File: train.py [hunk 1]
```diff
@@ -524,12 +517,33 @@ def __save_model(save_path, unet, text_encoder, tokenizer, scheduler, vae, save_
         if args.lr is not None:
             curr_lr = args.lr
             logging.info(f"Overriding LR from optimizer config with main config/cli LR setting: {curr_lr}")
+
+        text_encoder_lr_scale = optimizer_config.get("text_encoder_lr_scale", text_encoder_lr_scale)
+        if text_encoder_lr_scale != 1.0:
+            print(f" * Using text encoder LR scale {text_encoder_lr_scale}")
```
Comment (victorhallsai) [inline train.py:523]: should probably use logging.info here

File: train.py [hunk 2]
```diff
@@ -681,7 +696,7 @@ def sigterm_handler(signum, frame):
     )
     logging.info(f" Grad scaler enabled: {scaler.is_enabled()} (amp mode)")
 
-    epoch_pbar = tqdm(range(args.max_epochs), position=0, leave=True)
+    epoch_pbar = tqdm(range(args.max_epochs), position=0, leave=True, dynamic_ncols=True)
```
Comment (victorhallsai) [inline train.py:699]: are you able to test this in colab and jupyter (runpod/vast)?  the pbars are touchy, I think qslug finally got them working  in vast/runpod, and don't want to break anything

Comment (victorhallsai) [reply train.py:699]: colab pbars are already kinda broke but better not to make it worse

**Comment by victorhallsai** [comment]
Nice work! should be interesting to play with.

File: train.py (hunk 1 above)
Comment (damian0815) [reply train.py:523]: good call, thanks

File: train.py (hunk 2 above)
Comment (damian0815) [reply train.py:699]: i'll give it a shot

Comment (damian0815) [reply train.py:699]: seems to work fine, except for the ANSI color codes, but that isn't anything i've touched:

Screen Shot 2023-03-03 at 09 50 41

**Comment by damian0815** [comment]
@victorhallsai i took the liberty of defaulting sampling progress bars to `true` since they now clean up after themselves and no longer leave a big mess - following is what the text log looks like even after 21 rounds of validation and 4 sets of sample generation:
Screen Shot 2023-03-03 at 09 55 00

**Pull request merged**